server/src/main/java/org/elasticsearch/persistent/CompletionPersistentTaskAction.java

```diff
@@ -53,27 +56,45 @@ private CompletionPersistentTaskAction() {
 
         private long allocationId = -1;
 
+        private String localAbortReason;
+
         public Request() {}
 
         public Request(StreamInput in) throws IOException {
             super(in);
             taskId = in.readString();
             allocationId = in.readLong();
             exception = in.readException();
+            if (in.getVersion().onOrAfter(LOCAL_ABORT_AVAILABLE_VERSION)) {
+                localAbortReason = in.readOptionalString();
+            }
         }
 
-        public Request(String taskId, long allocationId, Exception exception) {
+        public Request(String taskId, long allocationId, Exception exception, String localAbortReason) {
             this.taskId = taskId;
             this.exception = exception;
             this.allocationId = allocationId;
+            this.localAbortReason = localAbortReason;
         }
 
         @Override
         public void writeTo(StreamOutput out) throws IOException {
+            if (localAbortReason != null && out.getVersion().before(LOCAL_ABORT_AVAILABLE_VERSION)) {
+                // This case cannot be handled by simply not serializing the new field, as the
+                // old master node would then treat it as a signal that the task should NOT be
+                // reassigned to a different node, i.e. completely different semantics. We
+                // should never get here in reality, as this action is for internal use only
+                // (it has no REST layer) and the places where it's called defend against this
+                // situation.
+                throw new IOException("attempt to abort a persistent task locally in a cluster that contains a node that is too old");
```

Comment:
nit: I prefer an IllegalArgumentException here - mainly for consistency, but also because it signals something done wrong by client (which is not really important here). We do that in other cases like `GetSnapshotsRequest`.